ral-facilities/scigateway
# Fixing login token bugs #735

## Description
This fixes issues which, through certain scenarios, would cause the user to be logged out of SciGateway when they shouldn't be. What would happen is the user's login session would be unexpectedly invalidated (their token in localStorage would be removed) which would cause errors when they try to navigate plugins. This was specifically a problem for the anon authenticator, which assigns a token automatically on initial page load; this bug would cause the token to be invalidated and errors would occur when the user navigated to plugins.

The issue was determined to be caused by two things:
1. when navigating to the login page and changing the authenticator selection from the dropdown but not actually logging in
2. when a login attempt failed

These changes now prevent this from happening and ensure that the autoLogin anon session remains valid. A further change has also been made which prevents a logged-in user from navigating to `/login` at all (previously the user could access it through direct navigation via the address bar, for example). Now what happens is the logged-in user will be redirected from `/login` to the homepage. The only way to access /login again is to explicitly log out (or be explicitly logged out, e.g. token expiry). The user can tell if they are logged out by looking for the 'Sign in' button at the top right of the page. Overall, this brings SciGateway in line with how other major websites typically handle login sessions - you are not allowed to login or attempt to login to an account while you are already logged in.

## Testing instructions
Load SciGateway with at least one plugin, ideally Dataview. Test with different authenticators, I used https://datagateway-dev.isis.stfc.ac.uk/api for my `authUrl` as it provides multiple authenticators. Keep an eye on the site's localStorage key-value pairs so you can see exactly when `scigateway:token` is being added, changed and removed. Check a token is assigned on initial page load and that it stays there while navigating different pages. Try changing the login authenticators and signing in using invalid credentials. The token should still be there. Also try logging in successfully to make sure the token changes. Do the same tests with this token to ensure it doesn't get removed unexpectedly. Hopefully you shouldn't see any errors when navigating the site!

Also try to navigate to `/login` while logged in. You should be redirected to the homepage every time. Then logout and make sure you can access `/login` again.


## Agile board tracking
Closes #735
